### PR TITLE
feat(server): add dynamic value substitution for prompts and tools

### PR DESCRIPTION
**Pull request by rgthelen**
## Summary

- Adds Jinja2-style template syntax (`{{variable | default}}`) for dynamic value substitution in prompts and tools
- Variables passed via new `variables` field in ChatRequest (OpenAI-compatible extension)
- Resolution happens at request time, maintaining full backwards compatibility

## Changes

- **New**: `server/services/template_service.py` - Core template engine with regex-based resolution
- **Modified**: `ChatRequest` in `projects.py` - Added optional `variables` field
- **Modified**: `PromptService` - Integrated template resolution into prompt merging
- **Modified**: `ChatOrchestratorAgent` - Added methods for runtime prompt/tool variable updates
- **New**: 124 unit tests across 6 test files
- **New**: 4 example projects in `examples/dynamic_values_*/`

## Test plan

- [x] Run unit tests: `cd server && uv run pytest tests/test_template_engine.py tests/test_chat_request_variables.py tests/test_prompt_template_resolution.py tests/test_tool_template_resolution.py tests/test_dynamic_values_integration.py tests/test_dynamic_values_edge_cases.py -v` (124 tests pass)
- [x] Live server test with dynamic variables
- [x] Verified defaults work when variables not provided
- [x] Verified backwards compatibility (existing configs work unchanged)

___

### **PR Type**
Enhancement

___

### **Description**
- Adds Jinja2-style template syntax (`{{variable | default}}`) for dynamic value substitution in prompts and tools

- Variables passed via new optional `variables` field in `ChatRequest` (OpenAI-compatible extension)

- Implements `TemplateService` with regex-based variable resolution supporting required and optional variables

- Integrates template resolution into `PromptService` and `ChatOrchestratorAgent` for runtime prompt/tool updates

- Includes 124 comprehensive unit tests across 6 test files and 4 example projects demonstrating usage

___

### Diagram Walkthrough

```mermaid
flowchart LR
  ChatRequest["ChatRequest<br/>+ variables field"]
  TemplateService["TemplateService<br/>resolve/resolve_object"]
  PromptService["PromptService<br/>resolve_prompts_for_model"]
  ChatOrchestrator["ChatOrchestratorAgent<br/>update_prompts/tools"]
  
  ChatRequest -->|variables| TemplateService
  TemplateService -->|resolved values| PromptService
  TemplateService -->|resolved values| ChatOrchestrator
  PromptService -->|resolved prompts| ChatOrchestrator
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>template_service.py</strong><dd><code>New template engine for variable substitution</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-94e5cda84dbaaf5264d413f11c27637c48409a584f9f8d4d63641ec9cdaafb88">+186/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>projects.py</strong><dd><code>Add variables field to ChatRequest model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-ab4cacdb0344ceca710f889b0d839b684ea3bd780df3f6125661cf30950b287a">+23/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>prompt_service.py</strong><dd><code>Integrate template resolution into prompt merging</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-e524139a47a8d63623875a8092faafa6b258e827d142dd5d59865f62882d20a5">+69/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>chat_orchestrator.py</strong><dd><code>Add methods for runtime prompt/tool variable updates</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-61c784391d5a45c2c6e79b1fd43056b67c8bcbf9d0558f3a8a00e8d4258517a8">+115/-3</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>test_template_engine.py</strong><dd><code>Unit tests for template service core functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-a8bf7f394615217ebc7610188c1691a1def5fdb3906006866f6e1d968d7ba4b0">+291/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_chat_request_variables.py</strong><dd><code>Tests for ChatRequest variables field acceptance</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-74a4bbc43ebe22825f09f1eccdc924579001301b33d6e947edf8ff3c2a0a91af">+163/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_prompt_template_resolution.py</strong><dd><code>Tests for prompt variable resolution integration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-de33556765a388172854cde559f89a86ee01ba9330b7670d9a416a98d6c63478">+406/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_tool_template_resolution.py</strong><dd><code>Tests for tool definition variable resolution</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-f510169a8a892873cd21b2a1209556f3a1f85563cf3dc1dbe65f787ed19d9e40">+295/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_dynamic_values_integration.py</strong><dd><code>Integration tests for prompts and tools together</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-8f8e455a5c777e12f0345af47f3fe8d6b8ccf5fa58dfff728d5ab6514e6af132">+300/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_dynamic_values_edge_cases.py</strong><dd><code>Edge case and error handling tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-119881ceba64df701074bbdbc76d91493ae57c4c0f234ae98e6ba46b2560a779">+245/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td><strong>test_template.py</strong><dd><code>Demo script showing template service capabilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-a666d844b3238db5e2d3ec471da10e8bfe7e75671ed10f1d920df511a3947c41">+140/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_api_request.sh</strong><dd><code>Bash demo for ChatRequest variables field</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-6a376ea31ecdf53ef140c51fcbe6363a32cc81cbcd692593fde6f3f995352263">+87/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>llamafarm.yaml</strong><dd><code>Example project with dynamic prompt variables</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-f5ef325c700e37de5c34d9e092978001d107bef120be0e10f00b046879ca3874">+30/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>demo.sh</strong><dd><code>Demo script for dynamic prompt variables</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-fd5cf989793e560e8964bcd68add94d0242abcf9f6a09efb40b0a5852771d97d">+101/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>llamafarm.yaml</strong><dd><code>Example project with dynamic tool variables</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-d47d7a52711c8e7405d02b0684e13d9b551d15948208fb1590567b5ce7cf7c39">+37/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>demo.sh</strong><dd><code>Demo script for dynamic tool variables</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-c96ef45ddd32649fcb95733d4b2620fa971be96b5dd7cb6d37e3dda36723af81">+118/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>llamafarm.yaml</strong><dd><code>Complete example with prompts and tools</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-aee5bb62ce33f81ac19407ec214468a746b8546baf08dfe2c36a385dc1d7cf6b">+75/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Documentation for complete dynamic values example</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-d888b3a553d49ab04e212aec1b1c24e2c35327e23b5894af62080b16a2f80aa8">+68/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>demo.sh</strong><dd><code>Comprehensive demo script for all features</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/672/files#diff-4d470025f11d6b0ba4d8452820a958b580d5b9714d3affbd174b58a0239b2235">+196/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by rgthelen** [comment]
## Response to Review Feedback

Thank you for the detailed review! Here's my response to the key points:

### Security Concerns

**System prompt override / Prompt injection / Template injection risks:**

These are **by design** and documented. The trust model is:
1. **Templates** come from config files (YAML in the project directory) - these are trusted
2. **Variable values** come from API request body - these are from API consumers

This is similar to other template systems (Jinja2, Handlebars) - the template is trusted, values are substituted. The API consumer passing malicious variable values could also just pass malicious prompt content directly in the `messages` array.

I've added a Security Note in the `TemplateService` docstring explicitly documenting this trust model.

**Variable logging risk:**

Only variable **names** are logged (not values) at DEBUG level. This helps debugging without exposing sensitive data. Production deployments should use INFO level or higher.

### Code Suggestions

**1. Use Jinja2 library:**
I considered this but chose a custom regex solution because:
- Jinja2 adds a 2MB dependency for a simple substitution feature
- The current implementation is ~50 lines and does exactly what we need
- Jinja2's advanced features (loops, conditionals, filters) aren't needed and would increase attack surface

**2. Preserve all prompt roles:**
Intentional - `update_prompts_with_variables()` only updates **system prompts** because those are the only prompts that come from config. User/assistant messages come from the conversation history, not config templates.

**3. Fix regex for default values:**
The current behavior is intentional - defaults containing `}` are supported (the test `test_default_with_pipe_in_value` verifies `{{var | a | b}}` results in default `a | b`). The regex correctly captures everything after the first `|` until `}}`.

**4. Preserve name field:**
The `PromptMessage` schema doesn't have a `name` field - this was based on an incorrect assumption in the review.

### Fixes Applied

- ✅ Added warning log when variables provided but agent doesn't support substitution
- ✅ Added Security Note documenting the trust model in `TemplateService` docstring

All 124 tests pass. CI should be green after this push.

**Comment by rgthelen** [comment]
## Additional Fixes Applied (98436567)

Based on the thorough code review comments, I've applied the following additional fixes:

### Error Handling (Fixed)
- ✅ Added proper `try-except` around template resolution in the API endpoint
- ✅ `TemplateError` now returns HTTP 400 with clear error message
- ✅ Added warning log before raising `TemplateError` for production debugging

### New Tests Added (10 new tests, 134 total)

**Security edge cases:**
- `test_potential_injection_in_value` - SQL injection patterns treated as literal strings
- `test_prompt_injection_in_value` - LLM prompt injection attempts treated literally  
- `test_template_markers_in_value` - Confirms no recursive resolution
- `test_very_long_value` - 100KB value works (no DOS via length)
- `test_default_with_single_closing_brace` - `{{var | {}}}` works correctly

**Type coercion:**
- `test_integer_value`, `test_float_value`, `test_boolean_value`
- `test_list_value`, `test_dict_value` - All convert to string repr

### Items Intentionally Not Changed

1. **Jinja2 library**: Custom regex is ~50 lines, doesn't add 2MB dependency, and limits attack surface
2. **Prompt role filtering**: Only system prompts come from config templates, so filtering is correct
3. **`None` → `"None"` behavior**: Documented, consistent with Python's `str()` semantics
4. **Default value trimming**: Intentional - leading/trailing whitespace in defaults is trimmed

All 134 tests pass. CI should be green.

**Comment by rgthelen** [comment]
## Response to Additional Review Suggestions

### 1. Use Jinja2 Library
**Decision: Keep custom implementation**

- Jinja2 adds ~2MB dependency for simple `{{var}}` and `{{var | default}}` substitution
- Jinja2's advanced features (loops, conditionals, filters, macros) increase attack surface
- Current implementation is ~50 lines, focused, and thoroughly tested (134+ tests)
- No feature gap - we don't need Jinja2's advanced templating

### 2. Preserve All Prompt Roles
**Decision: No change needed**

This is intentional behavior:
- `update_prompts_with_variables()` only updates **system prompts** because those are the only prompts that come from config templates
- User/assistant messages come from conversation history, not config templates
- The `system_prompt_generator.system_prompts` field is specifically for system prompts
- Putting all roles there would break the architecture

### 3. Fix Regex for `}` in Default Values
**Decision: Document limitation**

Tested the edge case:
- `{{var | {"key": "value"}}}` → Works correctly ✅
- `{{var | a}b}}` → Doesn't match (default containing `}` before `}}`)

This is an unlikely edge case (who puts `}` in a default value?). The suggested `(.*?)` fix would have its own issues. JSON defaults work fine, which is the realistic use case.

### 4. Preserve `name` Field in PromptMessage
**Decision: No change needed**

Verified the schema - `PromptMessage` class only has:
- `role: str`
- `content: str`
- `tool_call_id: Optional[str]`

There is no `name` field to preserve. The reviewer's suggestion was based on an incorrect assumption about the schema.

File: server/agents/chat_orchestrator.py [hunk 1]
```diff
@@ -96,6 +100,39 @@ def __init__(
         )
 
         super().__init__(config=config)
+        self._resolved_config_tools = None  # Initialize
```
Comment (BobbyRadford) [inline server/agents/chat_orchestrator.py]: this serves no purpose

File: server/agents/chat_orchestrator.py [hunk 2]
```diff
@@ -96,6 +100,39 @@ def __init__(
...
+
+    @property
+    def config_tools(self) -> list["ToolDefinition"]:
+        """Get config tools, using resolved version if available."""
+        from agents.base.types import ToolDefinition
```
Comment (BobbyRadford) [inline server/agents/chat_orchestrator.py]: move imports to the top of the file with the rest.

File: server/agents/chat_orchestrator.py [hunk 3]
```diff
@@ -61,6 +62,8 @@ class ChatOrchestratorAgent(LFAgent):
     _mcp_service: MCPService | None = None
     _mcp_tool_factory: MCPToolFactory | None = None
     _mcp_tools: list[type[BaseTool]] = []
+    _model_config: "Model"  # Store for later variable resolution
```
Comment (BobbyRadford) [inline server/agents/chat_orchestrator.py]: you should be able to declare the variable without setting a misleading "Model" value that does not reflect the actual data that will be stored in it in the initializer.

File: server/agents/chat_orchestrator.py [hunk 4]
```diff
@@ -61,6 +62,8 @@ class ChatOrchestratorAgent(LFAgent):
...
+    _resolved_config_tools: list["ToolDefinition"] | None = None  # Cache resolved tools
```
Comment (BobbyRadford) [inline server/agents/chat_orchestrator.py]: This comment is weird. this is not a "cache"

File: server/agents/chat_orchestrator.py [hunk 5]
```diff
@@ -76,6 +79,7 @@ def __init__(
 
         # Get the model config - if model_name is None, get_model returns the default
         model_config = ModelService.get_model(project_config, model_name)
+        self._model_config = model_config  # Store for variable resolution
```
Comment (BobbyRadford) [inline server/agents/chat_orchestrator.py]: This is an _unresolved_ model configuration value, i believe. As in, this does not have variable resolution applied yet. As such, this should be reflected in the variable name. For instance, `_model_config_unresolved`

File: server/api/routers/projects/projects.py [hunk 6]
```diff
@@ -532,8 +541,49 @@ async def chat(
         matched_docs = docs_service.match_docs_for_query(latest_user_message)
         agent.docs_context_provider.set_docs(matched_docs)
 
+    # Resolve template variables in prompts, config tools, and request tools
+    # Always resolve to apply defaults even when variables is empty/None
+    from core.logging import FastAPIStructLogger
+    from services.template_service import TemplateError, TemplateService
```
Comment (BobbyRadford) [inline server/api/routers/projects/projects.py]: move imports to the top of the file

File: server/api/routers/projects/projects.py [hunk 7]
```diff
@@ -532,8 +541,49 @@ async def chat(
...
+
+    template_logger = FastAPIStructLogger(__name__)
+
+    # Resolve prompts and config tools via agent methods
+    has_prompt_support = hasattr(agent, "update_prompts_with_variables")
+    has_tool_support = hasattr(agent, "update_config_tools_with_variables")
```
Comment (BobbyRadford) [inline server/api/routers/projects/projects.py]: these `hasattr` checks are superfluous.

**Comment by BobbyRadford** [comment]
@claude your proposal is mostly ok. however, you are not accounting for agent caching that happens at the api layer. if the variables are passed as an argument to the constructor, we would have to chec on each request for an agent in cache that matches those variables. that's weird. investigate this. i am fine if i'm wrong, but if that is a valid concern, you should consider an alternate approach. everything else is fine. stick to this area of concern only.

**Comment by rgthelen** [comment]
@BobbyRadford , I think I answered your concerns.  Claude keeps on coming up with more stuff.... happy to hit it up if you think it is worthwhile.

**Review by BobbyRadford — APPROVED** [review]
